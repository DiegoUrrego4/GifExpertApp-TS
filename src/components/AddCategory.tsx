import { ChangeEvent, useState } from "react";

interface Props {
	// setCategories: React.Dispatch<React.SetStateAction<string[]>>;
	onNewCategory: (value: string) => void;
}

export const AddCategory = ({ onNewCategory }: Props) => {
	const [inputValue, setInputValue] = useState("");

	const onInputChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
		setInputValue(target.value);
	};

	const onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
		event.preventDefault();
		if (inputValue.trim().length <= 1) {
			return;
		}
		onNewCategory(inputValue.trim());
		setInputValue("");
	};

	return (
		<form onSubmit={onSubmit}>
			<input type="text" placeholder="Buscar gifs" value={inputValue} onChange={onInputChange} />
		</form>
	);
};
