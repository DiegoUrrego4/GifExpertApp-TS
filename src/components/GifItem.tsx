interface Props {
	title: string;
	url: string;
}

export const GifItem = ({ title, url }: Props) => {
	return (
		<div className="grid__card">
			<div className="grid__card__image">
				<img src={url} alt={title} />
			</div>
			<div className="grid__card__title">
				<p>{title}</p>
			</div>
		</div>
	);
};
