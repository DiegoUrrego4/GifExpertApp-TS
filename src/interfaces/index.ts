export * from "./gif";
export * from "./Giphy-response";
