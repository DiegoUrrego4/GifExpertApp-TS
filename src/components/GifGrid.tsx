import { useEffect, useState } from "react";

import { getGifs } from "../helpers/getGifs";
import { Gif } from "../interfaces";
import { GifItem } from "./GifItem";

interface Props {
	category: string;
}

export const GifGrid = ({ category }: Props) => {
	const [images, setImages] = useState<Gif[]>([]);

	useEffect(() => {
		getGifs(category)
			.then((newImages: Gif[]) => {
				setImages(newImages);
			})
			.catch((err: string) => {
				throw new Error(`${err}`);
			});
	}, [category]);

	return (
		<>
			<h3>{category}</h3>
			<div className="card-grid">
				{images.map((image) => (
					<GifItem key={image.id} {...image} />
				))}
			</div>
		</>
	);
};
