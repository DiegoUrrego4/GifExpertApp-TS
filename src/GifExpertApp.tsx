import { useState } from "react";

import { AddCategory } from "./components/AddCategory";
import { GifGrid } from "./components/GifGrid";

export const GifExpertApp = () => {
	const [categories, setCategories] = useState(["One Punch"]);
	const existingCategories = categories.map((category) => category.toLowerCase());

	const onAddCategory = (newCategory: string): void => {
		if (existingCategories.includes(newCategory.toLowerCase())) {
			return;
		}
		setCategories([newCategory, ...categories]);
	};

	return (
		<>
			<h1>GifExpertApp</h1>
			<AddCategory onNewCategory={(value) => onAddCategory(value)} />
			<ol>
				{categories.map((category) => (
					<GifGrid key={category} category={category} />
				))}
			</ol>
		</>
	);
};
