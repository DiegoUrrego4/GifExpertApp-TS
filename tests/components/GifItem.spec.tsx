import { render, screen } from "@testing-library/react";

import { GifItem } from "../../src/components";

describe("Pruebas en <GifItem />", () => {
	const mockTitle = "Mock Title";
	const mockUrl = "https://my-image.gif/";

	test("should match with snapshot", () => {
		const { container } = render(<GifItem title={mockTitle} url={mockUrl} />);
		expect(container).toMatchSnapshot();
	});

	test("should show image with URL and ALT indicated", () => {
		render(<GifItem title={mockTitle} url={mockUrl} />);
		// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
		const { src, alt } = screen.getByRole("img") as HTMLImageElement;
		expect(src).toBe(mockUrl);
		expect(alt).toBe(mockTitle);
	});

	test("should shows title in component", () => {
		render(<GifItem title={mockTitle} url={mockUrl} />);
		expect(screen.getByText(mockTitle)).toBeTruthy();
	});
});
