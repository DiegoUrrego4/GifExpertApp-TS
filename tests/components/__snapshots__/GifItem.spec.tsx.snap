// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> should match with snapshot 1`] = `
<div>
  <div
    class="grid__card"
  >
    <div
      class="grid__card__image"
    >
      <img
        alt="Mock Title"
        src="https://my-image.gif/"
      />
    </div>
    <div
      class="grid__card__title"
    >
      <p>
        Mock Title
      </p>
    </div>
  </div>
</div>
`;
